source/volume_of_fluid/setup_initial_conditions.cc

```diff
@@ -0,0 +1,223 @@
+        sim.old_solution.block(volume_of_fluidN_blockidx) = sim.solution.block(volume_of_fluidN_blockidx);
+        sim.old_old_solution.block(volume_of_fluidN_blockidx) = sim.solution.block(volume_of_fluidN_blockidx);
+        sim.old_solution.block(volume_of_fluidLS_blockidx) = sim.solution.block(volume_of_fluidLS_blockidx);
+        sim.old_old_solution.block(volume_of_fluidLS_blockidx) = sim.solution.block(volume_of_fluidLS_blockidx);
+
+        // Update associated composition field
+        const typename Simulator<dim>::AdvectionField composition_field = Simulator<dim>::AdvectionField::composition(data[f].composition_index);
+        update_volume_of_fluid_composition (composition_field, data[f], sim.solution);
+        const unsigned int volume_of_fluid_C_blockidx = composition_field.block_index(this->introspection());
+        sim.old_solution.block(volume_of_fluid_C_blockidx) = sim.solution.block(volume_of_fluid_C_blockidx);
+        sim.old_old_solution.block(volume_of_fluid_C_blockidx) = sim.solution.block(volume_of_fluid_C_blockidx);
+      }
+  }
+
+  template <int dim>
+  void VolumeOfFluidHandler<dim>::initialize_from_composition_field (const VolumeOfFluidField<dim> field, const unsigned int f_ind)
+  {
+    LinearAlgebra::BlockVector initial_solution;
+
+    initial_solution.reinit(sim.system_rhs, false);
+
+    const QIterated<dim> quadrature (QMidpoint<1>(), n_init_samples);
+    FEValues<dim, dim> fe_init (this->get_mapping(), this->get_fe(), quadrature,
+                                update_JxW_values | update_quadrature_points);
+
+    std::vector<types::global_dof_index>
+    local_dof_indicies (this->get_fe().dofs_per_cell);
+
+    const FEVariable<dim> &volume_of_fluid_var = field.volume_fraction;
+    const unsigned int component_index = volume_of_fluid_var.first_component_index;
+    const unsigned int blockidx = volume_of_fluid_var.block_index;
+    const unsigned int volume_of_fluid_ind
+      = this->get_fe().component_to_system_index(component_index, 0);
+
+    // Initialize state based on provided function
+    for (auto cell : this->get_dof_handler().active_cell_iterators ())
+      {
+        if (!cell->is_locally_owned ())
+          continue;
+
+        // Calculate approximation for volume
+        cell->get_dof_indices (local_dof_indicies);
+
+        fe_init.reinit (cell);
+
+        double volume_of_fluid_val = 0.0;
+        double cell_vol = 0.0;
+
+        for (unsigned int i = 0; i < fe_init.n_quadrature_points; ++i)
+          {
+            double ptvolume_of_fluid = this->get_initial_composition_manager().initial_composition(fe_init.quadrature_point(i), f_ind);
+            volume_of_fluid_val += ptvolume_of_fluid * fe_init.JxW (i);
+            cell_vol += fe_init.JxW(i);
+          }
+
+        initial_solution (local_dof_indicies[volume_of_fluid_ind]) = volume_of_fluid_val/cell_vol;
+      }
+
+    initial_solution.compress(VectorOperation::insert);
+
+    sim.compute_current_constraints();
```

Comment:
could you add a comment here of the following sort:
```suggestion
    // Apply constraints and update solution blocks. This is duplicated from
    // Simulator<dim>::set_initial_temperature_and_compositional_fields()
    // in order to separate the volume of fluid algorithm as much as possible from
    // the rest of the code.
    sim.compute_current_constraints();
```

Also I think it would be more efficient to move all of the `compress` and following operations into the `set_initial_volume_fractions` function above. Currently you reapply the current constraints for *all* fields every time you initialize *one* field. Additionally you would eliminate the duplication of this code between the compositional field and level set initialization functions.